src/2024h2/sandboxed-build-script.md

```diff
@@ -0,0 +1,214 @@
+  Currently looking at [WebAssembly System Interface (WASI)](WASI) and [Cackle].
+* Provide a way to opt-in sandboxed build scripts for Cargo packages,
+  and design a configurable interface to grant permissions to each crate.
+* Based on the results of those experiments,
+  consider whether the implementation should be a third-party Cargo plugin first,
+  or make it into Cargo as an unstable feature (with a proper RFC).
+
+[wasi]: https://wasi.dev/
+[Cackle]: https://github.com/cackle-rs/cackle
+
+### The "shiny future" we are working towards
+
+These could become future goals if this one succeeds:
+
+* The sandboxed build script feature will be opted-in at first when stabilized.
+  By the next Edition, sandboxed build scripts will be on by default,
+  hardening the supply chain security.
+* Cargo users only need to learn one interface for both sandboxed proc-macros and build scripts.
+  The configuration for build scripts will also cover the needs for sandboxed proc-macros,
+* Crates.io and the [`cargo info`] command display the permission requirements of a crate,
+  helping developers choose packages based on different security level needs.
+* The runtime of the sandbox environment is swappable,
+  enabling the potential support of remote execution without waiting for a first-party solution.
+  It also opens a door to hermetic builds.
+* If we choose something like WebAssembly as the runtime of the sandbox environment,
+  the build script binaries could be pre-built and downloadable from user-defined pre-built binary servers.
+
+[`cargo info`]: https://crates.io/crates/cargo-information
+
+## Design axioms
+
+In order of importance, a sandboxed build script feature should provide the following properties:
+
+* **Restrict runtime file system and network access**, unless allowed explicitly.
+* **Easy to configure for using system (C) libraries.**
+  Probing and building from system libraries is the major use case of build scripts.
+  We should support it as a first-class citizen.
+* **Declarative configuration interface to grant permissions to packages.**
+  A declarative configuration helps us analyze permissions granted more easily,
+  without running the actual code.
+* **Don't block the build when the sandboxed feature is off.**
+  The crates.io ecosystem shouldn't rely on the interface to **successfully** build things.
+  That would hurt the integration with other external build systems.
+  It should work as if it is an extra layer of security scanning.
+* **Room for supporting different sandbox runtimes and strategies.**
+  This is for easier integration into external build systems,
+  as well as faster iteration for experimenting with new ideas.
+
+Currently out of scope:
+
+* Terminal user interface.
+* Pre-built build script binaries.
+* Hermetic builds, though this extension should be considered.
+* Support for all tier-1 and tier-2 with-host-tools platforms.
+  As an experiment, we follow what the chosen sandbox runtime provides us.
```

Comment:
How about "on-par build times".

We will no longer be able to share "host" and "target" builds when the host and target are the same because things we normally build for the host will be built for wasm.